File: backend/app/schemas/document.py
"""
Pydantic schemas for request validation and API response serialization.
"""
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# ─── Enums / Literals ────────────────────────────────────────────────────────

DocumentStatus = Literal["UPLOADED", "PROCESSING", "ANALYZED", "FAILED"]

DocumentType = Literal[
    "Rental Agreement",
    "Freelance Contract",
    "Terms of Service",
    "Other",
]

# ─── Upload ───────────────────────────────────────────────────────────────────


class UploadResponse(BaseModel):
    """Returned immediately after a successful file upload."""

    document_id: uuid.UUID
    filename: str
    document_type: DocumentType
    status: DocumentStatus


# ─── Document ─────────────────────────────────────────────────────────────────


class DocumentOut(BaseModel):
    """Full document metadata — returned by GET /api/documents/{id}."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    filename: str
    document_type: str
    file_type: str
    file_size: int
    status: DocumentStatus
    created_at: datetime
    updated_at: datetime


# ─── Document Pages ───────────────────────────────────────────────────────────


class DocumentPageOut(BaseModel):
    """Single extracted page."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    page_number: int
    text: str
    char_start: int | None
    char_end: int | None


class DocumentTextResponse(BaseModel):
    """Returned by GET /api/documents/{id}/text."""

    document_id: uuid.UUID
    total_pages: int
    pages: list[DocumentPageOut]


# ─── Health ───────────────────────────────────────────────────────────────────


class HealthResponse(BaseModel):
    status: str = Field(default="ok")

File: frontend/src/types/document.ts
// Document status enum matching backend
export type DocumentStatus = 'UPLOADED' | 'PROCESSING' | 'ANALYZED' | 'FAILED'

// Document type options
export type DocumentType =
  | 'Rental Agreement'
  | 'Freelance Contract'
  | 'Terms of Service'
  | 'Other'

export const DOCUMENT_TYPES: DocumentType[] = [
  'Rental Agreement',
  'Freelance Contract',
  'Terms of Service',
  'Other',
]

// Returned immediately after upload
export interface UploadResponse {
  document_id: string
  filename: string
  document_type: DocumentType
  status: DocumentStatus
}

// Full document metadata
export interface Document {
  id: string
  filename: string
  document_type: DocumentType
  file_type: string
  file_size: number
  status: DocumentStatus
  created_at: string
  updated_at: string
}

// A single extracted page
export interface DocumentPage {
  id: string
  document_id: string
  page_number: number
  text: string
  char_start: number | null
  char_end: number | null
}

// Response for GET /api/documents/{id}/text
export interface DocumentTextResponse {
  document_id: string
  total_pages: number
  pages: DocumentPage[]
}

// Upload state machine
export type UploadStage = 'idle' | 'uploading' | 'extracting' | 'completed' | 'failed'

export interface UploadState {
  stage: UploadStage
  progress: number // 0–100
  documentId?: string
  filename?: string
  error?: string
}
